data_warehouse/services/coverage_statistics_services.py

```diff
@@ -0,0 +1,194 @@
+                    technical_title=title))
+            low, high, mid = 0, len(CoverageStatisticsService.AGES) - 1, -1
+            while low <= high:
+                mid = (low + high) // 2
+                small = CoverageStatisticsService.AGES[mid][1][0]
+                big = CoverageStatisticsService.AGES[mid][1][1]
+                if small <= age <= big:
+                    break
+                elif age < small:
+                    high = mid - 1
+                else:
+                    low = mid + 1
+            age_range_label = CoverageStatisticsService.AGES[mid][0]
+            age_range = tuple(CoverageStatisticsService.AGES[mid][1])
+            if age_range_label in ages_groupby:
+                ages_groupby[age_range_label]['coverage_count'] += 1
+            else:
+                ages_groupby[age_range_label] = {}
+                ages_groupby[age_range_label]['coverage_count'] = 1
+                ages_groupby[age_range_label]['total_count'] = len(
+                    User.objects.filter(age__range=age_range))
+        data['titles'] = titles_groupby
+        data['ages'] = ages_groupby
+        data['departments'] = department_groupby
+        data['total'] = total_users_covered
+        return data
+
+    # pylint: disable=R0912
+    @staticmethod
+    def get_traning_records(user, program_id, department_id=None,
+                            start_time=None, end_time=None):
+        '''根据项目ID和起始时间查询所有的培训活动的所有的培训记录（包含校内和校外活动)
+
+        Parameters
+        ----
+        user: User
+            执行此操作的用户
+
+        program_ids: int
+            培训项目ID，为None或者-999表示查询所有的培训项目。
+
+        department_id: int
+            需要导出培训项目所属的部门，为None则导出当前用户有权管理的院系的所有培训项目。
+
+        start_time: datetime
+            培训活动开始时间
+
+        end_time: datetime
+            培训活动结束时间
+
+        Return
+        ------
+
+        record: queryset
+            包含Record的查询集
+        '''
+        department_ids = []
+        if user is None or not (user.is_department_admin or user
+                                .is_school_admin):
+            raise BadRequest('你不是管理员，无权操作。')
+        else:
+            if department_id is not None:
+                try:
+                    department = Department.objects.get(id=department_id)
+                except ObjectDoesNotExist:
+                    department = None
+                if department is None:
+                    raise BadRequest('部门不存在。')
+                if not user.check_department_admin(department):
+                    raise BadRequest('你不是该院系的管理员，无权操作。')
+                department_ids.append(department_id)
+            else:
+                if user.is_school_admin:
+                    department_ids.extend([
+                        dept.id for dept in Department.objects.all()])
+                else:
+                    department_ids.append(user.department.id)
+        assert department_ids
+        if start_time is None or end_time is None:
```

Comment:
```python
if end_time is None:
    end_time = now()
if start_time is None:
    start_time = end_time.replace(...)
```
Your current logic will override user's input when user specifies only one of these two variables.